analysis/R/decode_test.R

```diff
@@ -0,0 +1,299 @@
+  #   total: number of reports
+  results <<- c(estimates = list(), stds = list())
+
+  RunMultipleTests(title, TestEstimatesAndStdsHelper, repetitions,
+                   params, map, pdf, total)
+
+  ave_e <- apply(results$estimates,1:2, mean)
+  observed_stds <- apply(results$estimates,1:2, sd)
+  ave_stds <- apply(results$stds,1:2, mean)
+
+  ground_truth <- matrix(map %*% pdf, nrow = params$m, byrow = TRUE)
+
+  checkTrue(!any(abs(ave_e - ground_truth) > 1E-9 +  # tolerance level
+                                             (ave_stds / repetitions^.5) * 5),
+              "Averages deviate too much from expectations.")
+
+  checkTrue(!any(observed_stds > ave_stds * (1 + 5 * repetitions^.5)),
+            "Expected standard deviations are too high")
+
+  checkTrue(!any(observed_stds < ave_stds * (1 - 5 * repetitions^.5)),
+            "Expected standard deviations are too low")
+}
+
+TestEstimateBloomCounts <- function() {
+  # Unit tests for the EstimateBloomCounts function.
+
+  report4x2 <- list(k = 4, m = 2)  # 2 cohorts, 4 bits each
+  map0 <- Matrix(0, nrow = 8, ncol = 3, sparse = TRUE)  # 3 possible values
+  map0[1,] <- c(1, 0, 0)
+  map0[2,] <- c(0, 1, 0)
+  map0[3,] <- c(0, 0, 1)
+  map0[4,] <- c(1, 1, 1)  # 4th bit of the first cohort gets signal from all
+  map0[5,] <- c(0, 0, 1)  # 1st bit of the second cohort gets signal from v3
+
+  colnames(map0) <- c('v1', 'v2', 'v3')
+
+  pdf0 <- c(1/2, 1/3, 1/6)
+  names(pdf0) <- colnames(map0)
+
+  noise0 <- list(p = 0, q = 1, f = 0)  # no noise at all
+
+  TestEstimatesAndStds(repetitions = 1000, "Testing estimates and stds (1/3)",
+                       c(report4x2, noise0), map0, pdf0, 100)
+
+  noise1 <- list(p = 0.4, q = .6, f = 0.5)
+  TestEstimatesAndStds(repetitions = 1000, "Testing estimates and stds (2/3)",
+                       c(report4x2, noise1), map0, pdf0, 100)
+
+  # MEDIUM TEST: 100 values, 32 cohorts, 8 bits each, 10^6 reports
+  values <- 100
+
+  report8x32 <- list(k = 8, m = 32)  # 32 cohorts, 8 bits each
+
+  map1 <- matrix(rbinom(32 * 8 * values, 1, .25), nrow = 32 * 8, ncol = values)
+
+  colnames(map1) <- sprintf("v%d", 1:values)
+
+  pdf1 <- ComputePdf("zipf1", values)
+
+  TestEstimatesAndStds(repetitions = 100, "Testing estimates and stds (3/3)",
+                       c(report8x32, noise1), map1, pdf1, 10^9)
+}
+
+TestDecodeHelper <- function(params, map, pdf, num_clients,
+                             tolerance_l1, tolerance_linf) {
+  # Helper function for TestDecode. Simulates a RAPPOR run and checks results of
+  # Decode's output against the ground truth. Results are appended to a global
+  # list.
+
+  partition <- RandomPartition(num_clients, pdf)
+  counts <- GenerateCounts(params, map, partition, 1)
+  total <- sum(partition)
+
+  decoded <- Decode(counts, map, params)
+
+  decoded_partition <- setNames(decoded$fit$estimate, decoded$fit$strings)
+
+  results$estimates <<- MatrixVectorMerge(results$estimates, decoded_partition)
```

Comment:
I think this use of <<- is too subtle.  You can either pass the vars you need explicitly, or if there are too many of them, pass a single list() bundling them all.